### ISSUE TITLE
Disintegrate is emitting bindings out of order

### ISSUE THREAD

**Issue by zaxtax** [hakaru-dev/hakaru#30]
The following modified form of tugofwar:

```
def pulls(strength real):
    normal(strength, 1)

def winner(a real, b real):
    a_pull <~ pulls(a)
    b_pull <~ pulls(b)
    return (a_pull - b_pull)

alice <~ normal(0,1)
bob   <~ normal(0,1)
carol <~ normal(0,1)

match1 <~ winner(alice, bob)
match2 <~ winner(bob, carol)
match3 <~ winner(alice, carol)

return ((match1, match2), match3)
```

Disintegrates to the following invalid program

```
fn x13 pair(real, real): 
 (match x13: 
   (x30, x31): 
    x42 <~ normal(strength, 1)
    x55 <~ normal(strength, 1)
    pulls = fn strength real: normal(strength, 1)
    winner = fn a real: 
              fn b real: 
               a_pull <~ pulls(a)
               b_pull <~ pulls(b)
               return (a_pull - b_pull)
    alice <~ normal(0, 1)
    bob <~ normal(0, 1)
    carol <~ normal(0, 1)
    x59 <~ weight((exp((negate(((x30 + x42 - alice) ^ 2)) / 2))
                    / 
                   1
                    / 
                   sqrt((2 * pi))),
                  return ())
    x58 <~ weight((exp((negate(((x31 + x55 - bob) ^ 2)) / 2))
                    / 
                   1
                    / 
                   sqrt((2 * pi))),
                  return ())
    match3 <~ winner(carol, alice)
    return match3
   _: reject. measure(real))
```

Any ideas why?

**Comment by JacquesCarette** [comment]
This seems rather bad -- is still an issue?  Can we make it into a test case?

**Comment by zaxtax** [comment]
I think it is still an issue.

**Comment by JacquesCarette** [comment]
So who should be looking into it?  It seems to matter, as it is an example used by some of our 'competitiors'.

**Comment by zaxtax** [comment]
Totally, any help would be welcome!

**Comment by pravnar** [comment]
Here is a smaller example that exhibits the same error:

    def pulls(strength real):
        normal(strength, 1)
    
    def winner(a real, b real):
        a_pull <~ pulls(a)
        b_pull <~ pulls(b)
        return (a_pull - b_pull)
    
    alice <~ normal(0,1)
    bob   <~ normal(0,1)
    
    match1 <~ winner(alice, bob)
    
    return (match1, match1)

I'm adding it to Tests.Disintegrate. I can look into it.

**Comment by pravnar** [comment]
We fix this by going under "Neutral" terms (c420a10).
Disintegrating `examples/tugofwar.hk` (f05b829) produces the following program:

     fn x13 pair(real, real):
      (match x13:
        (x30, x31):
         x42 <~ normal(0, 1)
         x43 <~ normal(x42, 1)
         x56 <~ normal(0, 1)
         x57 <~ normal(x56, 1)
         pulls = fn strength real: normal(strength, 1)
         winner = fn a real:
                   fn b real:
                    a_pull <~ pulls(a)
                    b_pull <~ pulls(b)
                    return (a_pull - b_pull)
         alice <~ normal(0, 1)
         x61 <~ weight((exp((negate(((x30 + x43 - alice) ^ 2)) / 2))
                         /
                        1
                         /
                        sqrt((2 * pi))),
                       return ())
         x60 <~ weight((exp((negate(((x31 + x57 - x42) ^ 2)) / 2))
                         /
                        1
                         /
                        sqrt((2 * pi))),
                       return ())
         match3 <~ winner(alice, x56)
         return match3
        _: reject. measure(real))
